File: Android/ChannelViewer/app/src/main/java/com/phenixrts/demo/simpledemoapps/subsciber/subscribe/PhenixSubscribeComponent.java
/*
 * Copyright 2019 Phenix Real Time Solutions, Inc. Confidential and Proprietary. All Rights Reserved.
 *
 *  By using this code you agree to the Phenix Terms of Service found online here:
 *  http://phenixrts.com/terms-of-service.html
 */
package com.phenixrts.demo.simpledemoapps.subsciber.subscribe;

import static com.phenixrts.demo.simpledemoapps.subsciber.PhenixHelper.createChannelExpress;
import static com.phenixrts.system.Utilities.requireNonNull;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.SurfaceView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.phenixrts.express.ChannelExpress;
import com.phenixrts.express.ChannelExpressFactory;
import com.phenixrts.express.ExpressSubscriber;
import com.phenixrts.express.JoinChannelOptions;
import com.phenixrts.express.JoinChannelOptionsBuilder;
import com.phenixrts.express.JoinRoomOptions;
import com.phenixrts.express.RoomExpressFactory;
import com.phenixrts.pcast.AspectRatioMode;
import com.phenixrts.pcast.Renderer;
import com.phenixrts.pcast.RendererOptions;
import com.phenixrts.pcast.android.AndroidVideoRenderSurface;
import com.phenixrts.room.RoomService;
import java.io.IOException;

public class PhenixSubscribeComponent {

  private static final String TAG = PhenixSubscribeComponent.class.getSimpleName();

  private final Handler mainThreadHandler = new Handler(Looper.getMainLooper());

  private final ChannelExpress channelExpress;

  /**
   * PhenixHelper for easy subscribe to channels
   *
   * @param backendUri backend url to be used for creating {@link ChannelExpress}
   * @param pcastUri optional, for custom authentication
   */
  public PhenixSubscribeComponent(@NonNull Context context,
      @NonNull String backendUri,
      @Nullable String pcastUri) {

    Log.d(TAG, String.format("Creating ChannelExpress with backendUri [%s] and pcastUri [%s]",
        backendUri, pcastUri));

    this.channelExpress = createChannelExpress(requireNonNull(context),
        requireNonNull(backendUri), pcastUri);
  }

  /**
   * @param channelExpress to be used for all channel subscribe related operations
   */
  public PhenixSubscribeComponent(ChannelExpress channelExpress) {
    Log.d(TAG, "Reusing existing ChannelExpress");
    this.channelExpress = requireNonNull(channelExpress);
  }

  private JoinChannelOptions buildOptions(String channelAlias, SurfaceView playerView,
      SubscribeCapability... capabilities) {

    final String[] capabilityStrings = SubscribeCapability.convertCapabilities(capabilities);

    final JoinRoomOptions joinRoomOptions = RoomExpressFactory.createJoinRoomOptionsBuilder()
        .withRoomAlias(channelAlias)
        .withCapabilities(capabilityStrings)
        .buildJoinRoomOptions();

    JoinChannelOptionsBuilder joinChannelOptionsBuilder = ChannelExpressFactory
        .createJoinChannelOptionsBuilder()
        .withJoinRoomOptions(joinRoomOptions);

    // preview options
    if (playerView != null) {
      final AndroidVideoRenderSurface renderSurface = new AndroidVideoRenderSurface(
          playerView.getHolder());
      final RendererOptions rendererOptions = new RendererOptions();
      rendererOptions.aspectRatioMode = AspectRatioMode.LETTERBOX;

      joinChannelOptionsBuilder
          .withRenderer(renderSurface)
          .withRendererOptions(rendererOptions);
    }

    return joinChannelOptionsBuilder.buildJoinChannelOptions();
  }

  /**
   * Try to subscribe to channel and play it content, when ready. <br/> PLease provide listener and
   * save to local strong reference to {@link com.phenixrts.room.RoomService} and {@link
   * com.phenixrts.express.ExpressSubscriber} from {@link OnSubscribeCallback#onJoinChannelSuccess(RoomService)},
   * {@link OnSubscribeCallback#onSubscribeSuccess(ExpressSubscriber, Renderer)}. <br/> To
   * unsubscribe, use {@link RoomService#leaveRoom(RoomService.LeaveRoomCallback)}
   *
   * @param channelAlias of channel to be played
   * @param playerView surface to play video content on
   * @param callback listener for events.
   * @param capabilities subscribe configurations. If null - {@link SubscribeCapability#REAL_TIME}
   */
  public void subscribe(String channelAlias, @Nullable SurfaceView playerView,
      OnSubscribeCallback callback, SubscribeCapability... capabilities) {

    if (channelAlias == null || channelAlias.isEmpty() || callback == null) {
      throw new NullPointerException("Please provide configuration and listener");
    }

    channelExpress.joinChannel(buildOptions(channelAlias, playerView, capabilities),
        // join channel callback
        (requestStatus, roomService) ->
            mainThreadHandler.post(() -> {
              switch (requestStatus) {
                case OK:
                  callback.onJoinChannelSuccess(roomService);
                  break;
                default:
                  callback.onError(new SubscribeException.JoinChannelException(requestStatus));
              }
            }),
        // subscribe channel callback
        (requestStatus, expressSubscriber, renderer) ->
            mainThreadHandler.post(() -> {
              switch (requestStatus) {
                case OK:
                  callback.onSubscribeSuccess(expressSubscriber, renderer);
                  break;
                case NO_STREAM_PLAYING:
                  // We are joined channel, but there is no active content.
                  // We will automatically subscribe to channel, when content appear
                  callback.onNoActiveStream();
                  break;
                default:
                  callback.onError(new SubscribeException.SubscribeChannelException(requestStatus));
              }
            }));
  }

  /**
   * Completly stop subscription
   */
  public void stopSubscription(RoomService roomService, ExpressSubscriber expressSubscriber,
      Renderer renderer) {

    if (renderer != null) {
      renderer.stop();
    }

    if (expressSubscriber != null) {
      expressSubscriber.stop();
    }

    if (roomService != null) {
      try {
        roomService.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
  }

}

File: iOS/ChannelViewer/ChannelViewer/Subscriber/PhenixSubscribeComponent.swift
/*
 * Copyright 2019 Phenix Real Time Solutions, Inc. Confidential and Proprietary. All Rights Reserved.
 *
 * By using this code you agree to the Phenix Terms of Service found online here:
 * http://phenixrts.com/terms-of-service.html
 */

import UIKit
import PhenixSdk


/**
 * Class contains all logic to subscribe to a channel
 */
public class PhenixSubscribeComponent {

    public struct Configuration {
        var channelAlias: String
        // realtime is default
        var capability: SubscribeCapability?
    }

    private let channelExpress: PhenixChannelExpress

    init(channelExpress: PhenixChannelExpress) {
        self.channelExpress = channelExpress
    }

    /**
     * Prepare required configuration to subcscribe channel
     */
    private func prepareNativeConfiguration(configuration: Configuration, playerView: UIView!) -> PhenixJoinChannelOptions {
        let subscribeCapability = configuration.capability == nil ? [String]() : [configuration.capability!.rawValue]

        let joinRoomOptions = PhenixRoomExpressFactory.createJoinRoomOptionsBuilder()
            .withRoomAlias(configuration.channelAlias)
            .withRole(PhenixMemberRole.audience)
            .withCapabilities(subscribeCapability)
            .buildJoinRoomOptions()

        let renderLayer = playerView.layer

        let rendererOptions = PhenixRendererOptions()
        rendererOptions.aspectRatioMode = .letterbox

        let joinChannelOptions = PhenixChannelExpressFactory.createJoinChannelOptionsBuilder()
            .withJoinRoomOptions(joinRoomOptions)
            // Select the most recent stream. Viewing stream changes any time a new stream starts in the room.
            .withStreamSelectionStrategy(.mostRecent)
            // Alternatively you can use high availability if you are using multiple simultaneous publishers.
            // .withStreamSelectionStrategy(.highAvailability)
            .withRendererOptions(rendererOptions)
            .withRenderer(renderLayer)
            .buildJoinChannelOptions()

        return joinChannelOptions!
    }

    public func subscribe(configuration: Configuration, playerView: UIView, subscribeCallback: OnSubscribeCallback) {

        assert(Thread.current.isMainThread, "Needs to be called from main thread")

        let options = prepareNativeConfiguration(configuration: configuration, playerView: playerView)

        channelExpress.joinChannel(options, { (joinStatus, roomService) in // first of all, we should connect channel
            // Callbacks will be returned from worker thread.
            // For UI operations better to wrap it in mainthread
            DispatchQueue.main.async {
                switch(joinStatus) {
                case .ok: // we joined channel
                    subscribeCallback.onJoinChannelSuccess(roomService: roomService!)
                default: // something went wrong
                    subscribeCallback.onError(error: .joinChannel(status: joinStatus))
                }
            }
        }) { (suscribeStatus, subscriber, renderer) in // then we could subscribe to it conten
            // Callbacks will be returned from worker thread.
            // For UI operations better to wrap it in mainthread
            DispatchQueue.main.async {
                switch(suscribeStatus) {
                case .ok: // we subscribed.
                    subscribeCallback.onSubscribeSuccess(subscriber: subscriber!, renderer: renderer!)
                case .noStreamPlaying: // we subscribed, but there is no active content for playback
                    subscribeCallback.onNoActiveStream()
                default: // something went wrong
                    subscribeCallback.onError(error: .subscribeChannel(status: suscribeStatus))
                }
            }
        }
    }

}

public extension PhenixRequestStatus {
    public func getLocalisedStatusMessage() -> String {
        switch self {
        case .ok: return NSLocalizedString("PhenixRequestStatusOk", comment: "")
        case .noStreamPlaying: return NSLocalizedString("PhenixRequestStatusNoStreamPlaying", comment: "")
        case .badRequest: return NSLocalizedString("PhenixRequestStatusBadRequest", comment: "")
        case .unauthorized: return NSLocalizedString("PhenixRequestStatusUnauthorized", comment: "")
        case .conflict: return NSLocalizedString("PhenixRequestStatusConflict", comment: "")
        case .gone: return NSLocalizedString("PhenixRequestStatusGone", comment: "")
        case .notInitialized: return NSLocalizedString("PhenixRequestStatusNotInitialized", comment: "")
        case .notStarted: return NSLocalizedString("PhenixRequestStatusNotStarted", comment: "")
        case .upgradeRequired: return NSLocalizedString("PhenixRequestStatusUpgradeRequired", comment: "")
        case .failed: return NSLocalizedString("PhenixRequestStatusFailed", comment: "")
        case .capacity: return NSLocalizedString("PhenixRequestStatusCapacity", comment: "")
        case .timeout: return NSLocalizedString("PhenixRequestStatusTimeout", comment: "")
        case .notReady: return NSLocalizedString("PhenixRequestStatusNotReady", comment: "")
        }
    }
}
